### PR TITLE
[BD-20] Haystack Replacement and Elasticsearch upgrade for edx notes api

### PR DESCRIPTION
**Pull request by NikolayBorovenskiy**
**Description:** Haystack replacement and ES7 transition for edx-notes-api
Was done:
1. Run tests for Juniper
2. Update ES to version 7.
3. Run tests
4. Replace Haystack with elasticsearch-dsl, django-elasticsearch-dsl, django-elasticsearch-dsl-drf.
5. Update Tests

**Test coverage**
```
----------- coverage: platform linux, python 3.5.2-final-0 -----------
Name                                                   Stmts   Miss Branch BrPart  Cover
----------------------------------------------------------------------------------------
notesapi/__init__.py                                       0      0      0      0   100%
notesapi/urls.py                                           3      0      0      0   100%
notesapi/v1/__init__.py                                    0      0      0      0   100%
notesapi/v1/management/__init__.py                         0      0      0      0   100%
notesapi/v1/management/commands/__init__.py                0      0      0      0   100%
notesapi/v1/management/commands/bulk_create_notes.py      48     48     16      0     0%
notesapi/v1/migrations/0001_initial.py                     6      0      0      0   100%
notesapi/v1/migrations/0002_note_tags.py                   6      0      0      0   100%
notesapi/v1/migrations/0003_auto_20200703_1515.py          4      0      0      0   100%
notesapi/v1/migrations/__init__.py                         0      0      0      0   100%
notesapi/v1/models.py                                     25      1      6      1    94%
notesapi/v1/paginators.py                                  3      0      0      0   100%
notesapi/v1/permissions.py                                36      0     16      0   100%
notesapi/v1/search_indexes/__init__.py                     0      0      0      0   100%
notesapi/v1/search_indexes/backends/__init__.py            2      0      0      0   100%
notesapi/v1/search_indexes/backends/note.py                8      0      2      0   100%
notesapi/v1/search_indexes/documents/__init__.py           2      0      0      0   100%
notesapi/v1/search_indexes/documents/analyzers.py          4      0      0      0   100%
notesapi/v1/search_indexes/documents/note.py              28      1      2      0    97%
notesapi/v1/search_indexes/paginators.py                   4      0      0      0   100%
notesapi/v1/search_indexes/serializers/__init__.py         2      0      0      0   100%
notesapi/v1/search_indexes/serializers/note.py            22      0      8      0   100%
notesapi/v1/serializers.py                                15      0      0      0   100%
notesapi/v1/tests/__init__.py                              0      0      0      0   100%
notesapi/v1/tests/helpers.py                               7      0      0      0   100%
notesapi/v1/tests/test_models.py                          36      0      4      0   100%
notesapi/v1/tests/test_update_index.py                    34      0      2      0   100%
notesapi/v1/tests/test_views.py                          564      2     30      1    99%
notesapi/v1/urls.py                                        4      0      0      0   100%
notesapi/v1/utils.py                                       8      0      0      0   100%
notesapi/v1/views.py                                     162      5     38      6    94%
notesserver/__init__.py                                    0      0      0      0   100%
notesserver/docker_gunicorn_configuration.py              22     22      8      0     0%
notesserver/test_views.py                                 56      0      0      0   100%
notesserver/urls.py                                        3      0      0      0   100%
notesserver/views.py                                      51      0      8      1    98%
----------------------------------------------------------------------------------------
TOTAL                                                   1165     79    140      9    91%
```

**New index mapping for Elasticsearch7. Django model - Note**
```
{
  "notes_index_test": {
    "mappings": {
      "properties": {
        "course_id": {
          "type": "keyword"
        },
        "created": {
          "type": "date"
        },
        "id": {
          "type": "integer"
        },
        "quote": {
          "type": "text",
          "analyzer": "html_strip"
        },
        "ranges": {
          "type": "keyword"
        },
        "tags": {
          "type": "text",
          "analyzer": "case_insensitive_keyword"
        },
        "text": {
          "type": "text",
          "analyzer": "html_strip"
        },
        "updated": {
          "type": "date"
        },
        "usage_id": {
          "type": "keyword"
        },
        "user": {
          "type": "keyword"
        }
      }
    }
  }
}
``` 

**Reviewers:**
- @mikix 
- @dianakhuang 

**Post merge:**
- Delete working branch

### CODE REVIEW

**Comment by edx-webhook** [comment]
Thanks for the pull request, @NikolayBorovenskiy! I've created [BLENDED-445](https://openedx.atlassian.net/browse/BLENDED-445) to keep track of it in Jira.  More details are on the [BD-20](https://openedx.atlassian.net/wiki/spaces/COMM/pages/1596162490/BD-19+BD-20+Elasticsearch+Haystack+Upgrades) project page.

When this pull request is ready, tag your edX technical lead.

File: requirements/travis.txt [hunk 1]
```diff
@@ -8,7 +8,7 @@ appdirs==1.4.4            # via virtualenv
 certifi==2020.6.20        # via requests
 chardet==3.0.4            # via requests
 codecov==2.1.7            # via -r requirements/travis.in
-coverage==5.2             # via codecov
+coverage==5.1             # via codecov
```
Comment (mikix) [inline requirements/travis.txt:11]: A version downgrade?

File: notesserver/settings/yaml_config.py [hunk 2]
```diff
@@ -42,7 +42,7 @@
     DATABASES['default'][override] = value
 
 if ES_DISABLED:
-    HAYSTACK_CONNECTIONS = {}
-    INSTALLED_APPS.remove('haystack')
+    ELASTICSEARCH_DSL = {}
+    INSTALLED_APPS = [i for i in INSTALLED_APPS if i not in ES_APPS]
```
Comment (mikix) [inline notesserver/settings/yaml_config.py:46]: Don't we only add `ES_APPS` to `INSTALLED_APPS` if `not ES_DISABLED` anyway? In which case we wouldn't need this line, right? (Same question for other instance above in `test_es_disabled.py`)

File: notesapi/v1/search_indexes/backends/__init__.py [hunk 3]
```diff
@@ -0,0 +1,3 @@
+from .note import CompoundSearchFilterBackend
+
+__all__ = ('CompoundSearchFilterBackend',)
```
Comment (mikix) [inline notesapi/v1/search_indexes/backends/__init__.py:3]: Open edX style is usually to not define `__all__`, but doing it is fine. Just calling out that you don't need to.

File: notesapi/v1/utils.py [hunk 4]
```diff
@@ -0,0 +1,29 @@
+"""
+Utilities for notes api application
+"""
+
+from django.conf import settings
+from rest_framework.response import Response
+from rest_framework.utils.serializer_helpers import ReturnList
+
+
+class NotesPaginatorMixin:
```
Comment (mikix) [inline notesapi/v1/utils.py:10]: Nice mixin use.

File: notesapi/v1/search_indexes/backends/__init__.py (hunk 3 above)
Comment (NikolayBorovenskiy) [reply notesapi/v1/search_indexes/backends/__init__.py:3]: Thanks. I got it. I thought it was a good solution to explicitly define the interface of this module.

File: notesserver/settings/yaml_config.py (hunk 2 above)
Comment (NikolayBorovenskiy) [reply notesserver/settings/yaml_config.py:46]: Oh yes, you're right Michael. This line doesn't make any sense. Thanks for noticing this! 
I'll remove both here and in test_es_disabled.py

File: .travis/docker-compose-travis.yml [hunk 5]
```diff
@@ -12,8 +12,23 @@ services:
       MYSQL_PASSWORD: "secret"
       MYSQL_DATABASE: "edx_notes_api"
   es:
-    image: elasticsearch:1.5.2
+    image: docker.elastic.co/elasticsearch/elasticsearch:7.8.0
```
[the remark this answers is not in the document]
Comment (NikolayBorovenskiy) [reply .travis/docker-compose-travis.yml]: My logic was pretty simple. I opened the es docker documentation and got the docker image link there. Here is the link https://www.elastic.co/guide/en/elasticsearch/reference/current/docker.html
I am sure that the image from hub.docker.com will work for us too.

File: requirements/travis.txt (hunk 1 above)
Comment (NikolayBorovenskiy) [reply requirements/travis.txt]: This is pretty weird. This is either pip-compile tricks or something happened while resolving conflicts that were in dependencies. Anyway, there is now `coverage == 5.2`

**Comment by NikolayBorovenskiy** [comment]
@mikix thanks a lot for your review.
I left my answers under all your questions/comments. Pushed changes to the code where required.
I also fixed conflicts with the master branch and now PR is in good shape.

**Comment by regisb** [comment]
@NikolayBorovenskiy I stumbled upon this as I was working on the Koa upgrade: https://discuss.openedx.org/t/upgrading-tutor-for-koa/3499/7

Does this mean that the notes app no longer works with ES 1.5.2?

**Comment by dianakhuang** [comment]
@regisb i made a PR that should be 1.5.2 compatible for Koa, which hopefully resolves the confusion/difficulty: https://github.com/edx/edx-notes-api/pull/220

**Pull request merged**